src/core/file.cc

```diff
@@ -383,6 +383,40 @@ std::optional<directory_entry_type> dirent_type(const linux_dirent64& de) {
     return type;
 }
 
+#ifdef SEASTAR_COROUTINES_ENABLED
+coroutine::experimental::generator<directory_entry, dir_entry_buffer> make_list_directory_generator(coroutine::experimental::buffer_size_t buffer_size, thread_pool& pool, int fd) {
+    temporary_buffer<char> buf(8192);
+
+    while (true) {
+        auto ret = co_await pool.submit<syscall_result<long>>([fd, &buf] {
+            auto ret = ::syscall(__NR_getdents64, fd, reinterpret_cast<linux_dirent64*>(buf.get_write()), buf.size());
+            return wrap_syscall(ret);
```

Comment:
Can't it use the existing APIs? Why reimplement the syscall issue/parse?

Reply:
It's possible to make the new API appear on the file level only and it will re-use existing file_impl-s facilities. There's `make_list_directory_fallback_generator` in the last patch of this PR that does it. In the end of the day it needs seastar::queue on top what this implementation has.

Reply:
Yes, but why not call the existing API here? Instead of duplicating the implementation?

Reply:
@xemul i guess what Avi suggests is to make the generator a wrapper of the existing `list_directory()`?

Reply:
Yes, I got Avi's point, but my plan was to encourage file_impl::list_directory() switch to generator-style too. If implementing `make_list_directory_fallback_generator` on file level, it will keep all file_impls old-style for pretty much forever.